metrics/serve.go

```diff
@@ -15,39 +15,238 @@ package metrics
+	valGen *rand.Rand
+
+	gauges           []*prometheus.GaugeVec
+	counters         []*prometheus.CounterVec
+	histograms       []*prometheus.HistogramVec
+	nativeHistograms []*prometheus.HistogramVec
+	summaries        []*prometheus.SummaryVec
+
+	updateNotifyCh   chan struct{}
+	stopCh           chan struct{}
+	valueTick        *time.Ticker
+	seriesTick       *time.Ticker
+	metricTick       *time.Ticker
+	changeSeriesTick *time.Ticker
+
+	mu sync.Mutex
+}
+
+// NewCollector returns Prometheus collector that can be registered in registry
+// that handles metric creation and changes, based on the given configuration.
+func NewCollector(cfg Config) *Collector {
+	if cfg.GaugeMetricCount == 0 {
+		cfg.GaugeMetricCount = cfg.MetricCount // Handle deprecated field.
+	}
 
-func registerMetrics(metricCount, metricLength, metricCycle int, labelKeys []string) {
-	metrics = make([]*prometheus.GaugeVec, metricCount)
-	for idx := 0; idx < metricCount; idx++ {
-		gauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{
-			Name: fmt.Sprintf("avalanche_metric_%s_%v_%v", strings.Repeat("m", metricLength), metricCycle, idx),
-			Help: "A tasty metric morsel",
-		}, append([]string{"series_id", "cycle_id"}, labelKeys...))
-		promRegistry.MustRegister(gauge)
-		metrics[idx] = gauge
+	c := &Collector{
+		cfg:            cfg,
+		valGen:         rand.New(rand.NewSource(time.Now().UnixNano())),
+		updateNotifyCh: make(chan struct{}, 1),
+		stopCh:         make(chan struct{}),
 	}
+
+	if cfg.ValueInterval > 0 {
+		c.valueTick = time.NewTicker(time.Duration(cfg.ValueInterval) * time.Second)
+	}
+	if cfg.SeriesInterval > 0 {
+		c.seriesTick = time.NewTicker(time.Duration(cfg.SeriesInterval) * time.Second)
+	}
+	if cfg.MetricInterval > 0 {
+		c.metricTick = time.NewTicker(time.Duration(cfg.MetricInterval) * time.Second)
+	}
+	if cfg.SeriesChangeInterval > 0 {
+		c.changeSeriesTick = time.NewTicker(time.Duration(cfg.SeriesChangeInterval) * time.Second)
+	}
+	return c
 }
 
-func unregisterMetrics() {
-	for _, metric := range metrics {
-		promRegistry.Unregister(metric)
+func (c *Collector) UpdateNotifyCh() chan struct{} {
+	return c.updateNotifyCh
+}
+
+type Config struct {
+	MetricCount, GaugeMetricCount, CounterMetricCount, HistogramMetricCount, NativeHistogramMetricCount, SummaryMetricCount int
+	HistogramBuckets                                                                                                        int
+
+	LabelCount, SeriesCount        int
+	MaxSeriesCount, MinSeriesCount int
+	MetricLength, LabelLength      int
+
+	ValueInterval, SeriesInterval, MetricInterval, SeriesChangeInterval, SeriesChangeRate int
+
+	SpikeMultiplier     float64
+	SeriesOperationMode string
+	ConstLabels         []string
+}
+
+func NewConfigFromFlags(flagReg func(name, help string) *kingpin.FlagClause) *Config {
+	cfg := &Config{}
+	flagReg("metric-count", "Number of gauge metrics to serve. DEPRECATED use --gauge-metric-count instead").Default("0").
+		IntVar(&cfg.MetricCount)
```

Comment:
We can remove yes, just didn't want to introduce too many changes at once -- this might be the MOST used flag (most important one), so wanted to give ppl time to adapt.

This also reminded me I might want to walk through docs to update them if they use metric-count.

Let's take it easy - added issue about this https://github.com/prometheus-community/avalanche/issues/82